Model/CustomItemModel.php

```diff
@@ -95,33 +94,62 @@ public function save(CustomItem $customItem, bool $dryRun = false): CustomItem
         $customItem->setModifiedBy($user);
         $customItem->setModifiedByUser($user->getName());
         $customItem->setDateModified($now->getUtcDateTime());
-
-        if (!$dryRun) {
-            $this->entityManager->persist($customItem);
-        }
-
-        $customItem->getCustomFieldValues()->map(function (CustomFieldValueInterface $customFieldValue) use ($dryRun): void {
-            $this->customFieldValueModel->save($customFieldValue, $dryRun);
-        });
+        $customItem->setUniqueHash($customItem->createUniqueHash());
 
         $errors = $this->validator->validate($customItem);
 
         if ($errors->count() > 0) {
             throw new InvalidValueException($errors->get(0)->getMessage());
         }
 
-        $customItem->recordCustomFieldValueChanges();
-
-        $this->dispatcher->dispatch(CustomItemEvents::ON_CUSTOM_ITEM_PRE_SAVE, $event);
+        $this->dispatcher->dispatch(CustomItemEvents::ON_CUSTOM_ITEM_PRE_SAVE, new CustomItemEvent($customItem, $customItem->isNew()));
 
         if (!$dryRun) {
-            $this->entityManager->flush($customItem);
-            $this->dispatcher->dispatch(CustomItemEvents::ON_CUSTOM_ITEM_POST_SAVE, $event);
+            if ($customItem->isNew()) {
+                // Custom item is new so we need to upsert it to atomically find whether it exists based on unique fields or not.
+                $this->entityManager->detach($customItem);
+                $this->customItemRepository->upsert($customItem);
+
+                // We need to re-attach the entity to the entity manager so that it can be saved by the rest of the code.
+                $customFieldValues = $customItem->getCustomFieldValues();
+                $customItem        = $this->fetchEntity($customItem->getId());
+
+                foreach ($customFieldValues as $customFieldValue) {
+                    $customFieldValue->setCustomItem($customItem);
+                    $customItem->addCustomFieldValue($customFieldValue);
+                }
+            } else {
+                $this->entityManager->persist($customItem);
+                $this->entityManager->flush();
+            }
+
+            $customItem->getCustomFieldValues()->map(
+                fn (CustomFieldValueInterface $customFieldValue) => $this->customFieldValueModel->save($customFieldValue, $dryRun)
+            );
+
+            $customItem->recordCustomFieldValueChanges();
+
+            $this->dispatcher->dispatch(CustomItemEvents::ON_CUSTOM_ITEM_POST_SAVE, new CustomItemEvent($customItem, $customItem->isNew()));
         }
 
         return $customItem;
     }
 
+    /**
+     * @param object $entity
+     * @param null $extra
+     * @throws InvalidValueException
+     */
+    public function unlockEntity($entity, $extra = null): void
```

Comment:
the second parameter is not being used anywhere?

Reply:
Since this is overriding `app/bundles/CoreBundle/Model/FormModel.php::unlockEntity()`, this declaration should be compatible with that one. That's why the second parameter has been added.